File: app/models/track.py
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey
from sqlalchemy.sql import func
from app.models.database import Base
from sqlalchemy.dialects.postgresql import JSON


class Track(Base):
    __tablename__ = "tracks"
    id = Column(String, primary_key=True, index=True) # Unique track ID
    title = Column(String, nullable=False) # Track title
    artist_id = Column(String, ForeignKey("artists.id"), nullable=False)  # Artist reference
    album_id = Column(String, ForeignKey("albums.id"), nullable=False)  # Album reference
    duration_ms = Column(Integer, nullable=False) # Duration of each track
    popularity = Column(Integer, nullable=True) # Popularity value
    aoty_score = Column(Integer, nullable=True) # AOTY User Score
    play_count = Column(Integer, nullable=True) # Last.fm playcount 
    audio_features = Column(JSON, nullable=False) # JSON field for audio features (tempo, energy, etc.)
    preview_url = Column(String, nullable=True)  # Spotify preview URL
    cover_url = Column(String, nullable=True) # URL to song/album cover image
    source = Column(String, nullable=False) # Spotify/Last.fm
    


    

File: frontend/app/src/lib/types/track.ts
/**
 * Unified track interface for all recommendation sources
 * (Agent, ML, Spotify, Last.fm, AOTY, etc.)
 */
export interface UnifiedTrack {
  // Core identifiers
  id: string;
  spotify_id?: string;
  lastfm_id?: string;
  
  // Basic metadata
  name: string;
  artist: string;
  artists?: string[]; // Additional artists for collaborations
  album?: string;
  album_id?: string;
  
  // Media
  artwork_url?: string;
  preview_url?: string;
  
  // Links
  spotify_url?: string;
  lastfm_url?: string;
  youtube_url?: string;
  
  // Audio characteristics
  duration_ms?: number;
  explicit?: boolean;
  popularity?: number;
  release_date?: string;
  
  // Recommendation context
  source: RecommendationSource;
  similarity_score?: number;
  confidence_score?: number;
  recommendation_reason?: string;
  
  // Classification
  genres?: string[];
  tags?: string[];
  mood?: string[];
  
  // Audio features (for ML recommendations)
  audio_features?: {
    danceability?: number;
    energy?: number;
    valence?: number;
    acousticness?: number;
    instrumentalness?: number;
    liveness?: number;
    speechiness?: number;
    tempo?: number;
    key?: number;
    mode?: number;
    time_signature?: number;
  };
  
  // Metadata
  created_at?: string;
  updated_at?: string;
  metadata?: Record<string, any>;
}

export type RecommendationSource = 
  | 'agent'
  | 'ml_hybrid'
  | 'ml_collaborative'
  | 'ml_content'
  | 'spotify_search'
  | 'spotify_playlist'
  | 'spotify_recommendations'
  | 'lastfm_similar'
  | 'lastfm_user'
  | 'aoty_similar'
  | 'aoty_charts'
  | 'database'
  | 'fallback';

/**
 * Recommendation context for displaying groups of tracks
 */
export interface RecommendationSet {
  id: string;
  title: string;
  description?: string;
  tracks: UnifiedTrack[];
  source: RecommendationSource;
  confidence: number;
  created_at: string;
  metadata?: {
    query?: string;
    user_input?: string;
    tools_used?: string[];
    explanation?: string;
    [key: string]: any;
  };
}

/**
 * User feedback on recommendations
 */
export interface TrackFeedback {
  track_id: string;
  user_id: string;
  feedback_type: 'like' | 'dislike' | 'save' | 'skip' | 'play_full' | 'share';
  timestamp: string;
  context?: {
    source: RecommendationSource;
    recommendation_set_id?: string;
    listening_session_id?: string;
    [key: string]: any;
  };
}